### PR TITLE
[WIP] Spark Shuffle prototype

### PR DESCRIPTION
**Pull request by mbasmanova**
CC: @beroyfb

### CODE REVIEW

File: velox/exec/tests/PartitionAndSerialize.h [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  PartitionAndSerializeNode(
+      const core::PlanNodeId& id,
+      std::vector<core::TypedExprPtr> keys,
+      int numPartitions,
+      RowTypePtr outputType,
+      core::PlanNodePtr source)
+      : core::PlanNode(id),
+        keys_{std::move(keys)},
+        numPartitions_{numPartitions},
+        outputType_{std::move(outputType)},
```
Comment (beroyfb) [inline velox/exec/tests/PartitionAndSerialize.h:34]: Why does it need the outputType to be set from outside when it has to be ROW(p, d)

File: velox/exec/tests/PartitionAndSerialize.cpp [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+  void computePartitions(FlatVector<int32_t>& partitionsVector) {
+    auto numInput = input_->size();
+
+    partitions_.resize(numInput);
+    partitionFunction_->partition(*input_, partitions_);
+
+    // TODO Avoid copy.
+    partitionsVector.resize(numInput);
+    auto rawPartitions = partitionsVector.mutableRawValues();
+    std::memcpy(rawPartitions, partitions_.data(), sizeof(int32_t) * numInput);
```
Comment (beroyfb) [inline velox/exec/tests/PartitionAndSerialize.cpp:109]: Is there another way to write partitions into the raw vectors and avoid the vector copy? I think we can have another version of partition function to do that.

File: velox/exec/tests/PartitionAndSerialize.cpp [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    auto numInput = input_->size();
+
+    dataVector.resize(numInput);
+
+    // Compute row sizes.
+    rowSizes_.resize(numInput);
+
+    size_t totalSize = 0;
+    for (auto i = 0; i < numInput; ++i) {
+      size_t rowSize = velox::row::UnsafeRowDynamicSerializer::getSizeRow(
```
Comment (beroyfb) [inline velox/exec/tests/PartitionAndSerialize.cpp:122]: Use ```auto```

File: velox/exec/tests/PartitionAndSerialize.cpp [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    }
+
+    // Allocate memory.
+    auto buffer = dataVector.getBufferWithSpace(totalSize);
+
+    // Serialize rows.
+    auto rawBuffer = buffer->asMutable<char>();
+    size_t offset = 0;
+    for (auto i = 0; i < numInput; ++i) {
+      dataVector.setNoCopy(i, StringView(rawBuffer + offset, rowSizes_[i]));
```
Comment (beroyfb) [inline velox/exec/tests/PartitionAndSerialize.cpp:135]: Why do we need ```setNoCopy``` here?

File: velox/exec/tests/ShuffleWrite.cpp [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+                << std::endl;
+    }
+  }
+
+  RowVectorPtr getOutput() override {
+    return nullptr;
+  }
+
+  BlockingReason isBlocked(ContinueFuture* future) override {
+    return BlockingReason::kNotBlocked;
```
Comment (beroyfb) [inline velox/exec/tests/ShuffleWrite.cpp:55]: Why is that the case here?

File: velox/exec/tests/PartitionAndSerialize.h (hunk 1 above)
Comment (mbasmanova) [reply velox/exec/tests/PartitionAndSerialize.h:34]: The shape of the output type is fixed: INTEGER + VARBINARY, but the names of the individual fields are not. The output type allows the planner to choose symbol names in whatever way makes sense.

File: velox/exec/tests/PartitionAndSerialize.cpp (hunk 2 above)
Comment (mbasmanova) [reply velox/exec/tests/PartitionAndSerialize.cpp:109]: Yes, we can have an API that takes a pre-allocated buffer as uint32_t*

```
partitionFunction_->partition(*input_, rawPartitions);
```

File: velox/exec/tests/PartitionAndSerialize.cpp (hunk 4 above)
Comment (mbasmanova) [reply velox/exec/tests/PartitionAndSerialize.cpp:135]: FlatVector<StringView>::set() method copies the StringView into internal string buffers. We don't want this code, hence, using setNoCopy.

File: velox/exec/tests/ShuffleWrite.cpp (hunk 5 above)
Comment (mbasmanova) [reply velox/exec/tests/ShuffleWrite.cpp:55]: ShuffleWriteOperator is never blocked because it is not waiting for anything.

**Pull request closed without merging**